wagtail_footnotes/blocks.py

```diff
@@ -15,69 +16,93 @@ class RichTextBlockWithFootnotes(RichTextBlock):
         )
         template = get_template(template_name)
-        return template.render({"index": index})
+        return template.render({"index": index, "reference_index": reference_index})
 
     def replace_footnote_tags(self, value, html, context=None):
         if context is None:
             new_context = self.get_context(value)
         else:
             new_context = self.get_context(value, parent_context=dict(context))
 
-        if not isinstance(new_context.get("page"), Page):
+        page = new_context.get("page")
+        if page is None or not isinstance(page, Page):
             return html
 
-        page = new_context["page"]
-        if not hasattr(page, "footnotes_list"):
-            page.footnotes_list = []
-        self.footnotes = {
+        # Map Footnote UUIDs to Footnote instances to simplify lookups once a reference has been found in the text.
+        # NOTE: Footnotes may exist in the database for a given page but this does not necessarily mean that the
+        # footnote was referenced in the text.
+        self.all_footnotes = {
             str(footnote.uuid): footnote for footnote in page.footnotes.all()
         }
 
+        # Patch the page to track the footnotes that are actually referenced in the text, so that they can be rendered
+        # in footnotes.html
+        if not hasattr(page, "footnotes_list"):
+            page.footnotes_list = []
+
         def replace_tag(match):
+            footnote_uuid = match.group(1)
             try:
-                index = self.process_footnote(match.group(1), page)
-            except (KeyError, ValidationError):
+                footnote = self.attach_footnote_to_page(footnote_uuid, page)
+            except KeyError:
                 return ""
             else:
-                return self.render_footnote_tag(index)
+                # Add 1 to the footnote index as footnotes are rendered in footnotes.html using `{{ forloop.counter }}`
+                # which is 1-based.
+                footnote_index = page.footnotes_list.index(footnote) + 1
+                reference_index = footnote.references[-1]
+                # Supplying both indexes allows for unique id values to be generated in the HTML. E.g., the first
+                # reference to the first footnote will have `id="footnote-source-1-1"`, and the second reference to the
+                # first footnote will have `id="footnote-source-1-2"`, etc.
+                return self.render_footnote_tag(footnote_index, reference_index)
 
         # note: we return safe html
         return mark_safe(FIND_FOOTNOTE_TAG.sub(replace_tag, html))  # noqa: S308
 
     def render(self, value, context=None):
-        if not self.get_template(value=value, context=context):
-            return self.render_basic(value, context=context)
-
         html = super().render(value, context=context)
         return self.replace_footnote_tags(value, html, context=context)
 
-    def render_basic(self, value, context=None):
```

Comment:
I removed `render_basic` because `super.render()` will already return this, and at the end of the day, all that is happening is regex substitutions on the output, which is handled in `render()` above.